[DocumentIntelligence] Add SDK samples for add-on capabilities (#40603)
Azure/azure-sdk-for-net: sdk/documentintelligence/Azure.AI.DocumentIntelligence/samples/Sample_AddOnCapabilities.md

# Analyze a document with add-on capabilities

> This sample demonstrates how to analyze document with add-on capabilities, using a document as an example. For more information about add-on capabilities, see the [documentation][docint_addon].

Reviewer: ```suggestion
This sample demonstrates how to analyze a document with add-on capabilities. For more information about the supported features, see the [service documentation][docint_addon].

```

To get started you'll need a Cognitive Services resource or a Document Intelligence resource. See [README][README] for prerequisites and instructions.

Add-on capabilities are available within all models except for the Business card model. This sample uses Layout model ("prebuilt-layout") to demonstrate. When analyzing a document, you can specify a list of optional [analysis features][sdk_docfeature] to enable certain add-on capabilities.

The following capabilities are free:
- Barcodes
- Languages
- KeyValuePairs

The following capabilities will incur additional charges, see [pricing][docint_pricing]:
- Formulas
- OcrHighResolution

> - FontStyling

Reviewer: We usually avoid this kind of information (pricing, supported models) in the SDK docs since they tend to get outdated pretty quickly. It's preferable to link to service documentation so the information gets centralized in one place.

Author: Sure. Will remove them.

## Creating a `DocumentIntelligenceClient`

To create a new `DocumentIntelligenceClient` you need the endpoint and credentials from your resource. In the sample below you'll use a Document Intelligence API key credential by creating an `AzureKeyCredential` object that, if needed, will allow you to update the API key without creating a new client.

You can set `endpoint` and `apiKey` based on an environment variable, a configuration setting, or any way that works for your application.

```C# Snippet:CreateDocumentIntelligenceClient
string endpoint = "<endpoint>";
string apiKey = "<apiKey>";
var client = new DocumentIntelligenceClient(new Uri(endpoint), new AzureKeyCredential(apiKey));
```

## High resolution extraction

> To extract content from a given file at a URI with improved quality through the add-on high resolution capability, use the `AnalyzeDocumentAsync` method and specify `DocumentAnalysisFeature.OcrHighResolution` as the analysis features. The returned value is an `AnalyzeResult` object containing data about the submitted document.

Reviewer: I can understand other capabilities, but is there any point in having high resolution extraction as a sample if it doesn't change how we parse the output? I think we don't need to have a sample for every single add-on capability if we don't have anything special to show. Customers can figure out how to pass the `DocumentAnalysisFeature.OcrHighResolution` option with other samples.

Author: You're absolutely right. The output will maintain the same structure with this capability. However, our goal is to provide a sample corresponding to each supported capability. If the sample code is kind of duplicate with existing ones, can we still keep the description part but remove the code snippet?

Author: Removed the code snippets for high resolution.

Uri uriSource = new Uri("<uriSource>");

var content = new AnalyzeDocumentContent()
{
    UrlSource = uriSource
};

List<DocumentAnalysisFeature> features = new List<DocumentAnalysisFeature>();
features.Add(DocumentAnalysisFeature.StyleFont);

var operation = await client.AnalyzeDocumentAsync(WaitUntil.Completed, "prebuilt-layout", content, features: features);
AnalyzeResult result = operation.Value;

// DocumentStyle has the following font related attributes:
var similarFontFamilies = new Dictionary<string, List<DocumentStyle>>(); // e.g., 'Arial, sans-serif
var fontStyles = new Dictionary<FontStyle, List<DocumentStyle>>(); // e.g, 'italic'
var fontWeights = new Dictionary<FontWeight, List<DocumentStyle>>(); // e.g., 'bold'
var fontColors = new Dictionary<string, List<DocumentStyle>>(); // in '#rrggbb' hexadecimal format
var fontBackgroundColors = new Dictionary<string, List<DocumentStyle>>(); // in '#rrggbb' hexadecimal format

if (result.Styles.Any(s => s.IsHandwritten != null && s.IsHandwritten.Value == true))
{
    Console.WriteLine("Document contains handwritten content");
}
else
{
    Console.WriteLine("Document does not contain handwritten content");

> }

Reviewer: How is this helpful to customers? I guess they would want to extract what exactly is the handwritten content instead.

Author: Print the handwritten content instead.

Console.WriteLine("\n----Fonts styles detected in the document----");

// Iterate over the styles and group them by their font attributes.
foreach (var style in result.Styles)
{
    if (!string.IsNullOrEmpty(style.SimilarFontFamily))
    {
        if (similarFontFamilies.ContainsKey(style.SimilarFontFamily))
        {
            similarFontFamilies[style.SimilarFontFamily].Add(style);
        }
        else
        {
            similarFontFamilies.Add(style.SimilarFontFamily, new List<DocumentStyle>() { style });
        }
    }

>     if (style.FontStyle != null && style.FontStyle.HasValue)

Reviewer: As a struct, checking `FontStyle != null` and `FontStyle.HasValue` is the same thing. You can leave just one of the comparisons.

The same comment applies to other struct properties such as `FontWeight`.
